### PR TITLE
Redesign rule and violation

### PR DESCRIPTION
**Pull request by TheFirstGameruler**
The idea discussed for Rule and Violation. As well as the folder structure.

### CODE REVIEW

**Review by TheFirstGameruler — COMMENTED** [review]
working version is now here!!!

**Review by orlandoisepic — APPROVED** [review]
I think this looks good. Well done!

File: preciceconfigchecker/rule.py [hunk 1]
```diff
@@ -1,37 +1,74 @@
...
-    name:str
-    message:str
-    possible_solutions:str
-    followed:bool = True
-    numbers_not_followed:int = 0
-
-    def __init__(self, severity:Severity = Severity.INFO, name:str = "Rule", message:str = "", possible_solutions:str = "") -> None:
+        Args:
+            severity (Severity): Type
+            problem (str): Short explanation of what the rule is supposed to check in general.
```
[the remark this answers is not in the document]
Comment (TheFirstGameruler) [reply preciceconfigchecker/rule.py:18]: long live: copy, paste

The idea behind it: If you have several violations per rule, a short explanation at the beginning is helpful to understand the following violations more quickly. We can describe each individual violation in as much detail as we want.


If we only expect a small number of violations, or if the violations themselves briefly describe the problem, we can also make this parameter optional.

[the remark this answers is not in the document]
Comment (orlandoisepic) [reply preciceconfigchecker/rule.py:18]: I agree; what I meant was an operation in Rule, like print_info, which prints general info about the rule. Then we avoid having to initialize a Rule with a long string which explains why the rule has its name. Instead, we can use the method if we need the info

[the remark this answers is not in the document]
Comment (TheFirstGameruler) [reply preciceconfigchecker/rule.py:18]: But the explanation should be short.

[the remark this answers is not in the document]
Comment (TheFirstGameruler) [reply preciceconfigchecker/rule.py:18]: Oh yes, then we can set these directly in the overwritten class. Then you don't have to do this when initializing.

[the remark this answers is not in the document]
Comment (TheFirstGameruler) [reply preciceconfigchecker/rule.py]: Done

**Review by TheFirstGameruler — COMMENTED** [review]
'problem' and 'severity' are now abstract properties. No longer need to be passed to the constructor.

**Review by orlandoisepic — APPROVED** [review]
Looks good

**Pull request merged**